### ISSUE TITLE
Migrate monorepo to strict TS

### ISSUE THREAD

**Issue by kasperpeulen** [storybookjs/storybook#22176]
#### What
We want to step our TS game in the monorepo and enable strict typescript in all packages!

#### Why
Having TS track for you if a variable might be null or not, enables us to code with much more confidence, 
and also gives us quick in editor feedback, when you make assumptions that are not actually true! 

#### How

Feel free to contribute to any of packages in the list below!

After you cloned the repo, you can best run this command:
`yarn task --task check --no-link --start-from=install`

This will install all dependencies, compile it (including dts files, that is what no-link does), and then check typescript errors in all packages.

To enable strict mode you will need:
1. Change in `tsconfig.json` `"strict": false` -> `"strict": true`
2. Run `yarn check` in the directory of the package.
3. Fix all TS errors. 

We would like to change as little as possible of the actual runtime behavior in this migration.
However, we also don't want to simply silence the compiler in every error with `!`, `as` or `ts-ignore` to get this migration in.
As a rule of thumb, if the logic is easy enough, prefer improving the code (e.g. add a null check, throw when not null, possibly use https://github.com/alexreardon/tiny-invariant) over silencing the compiler.
If the change needed to do the right thing, is too risky, and not in your expertise, it is okay to silence the compiler.
It is not ideal, but we still gain the benefit that new code written will have extra typesafety.

| package name                                                                   | assign to        | merged | PR |
|------------------------------------------------------|-------------- |------------|----|
|@storybook/addon-backgrounds| @kasperpeulen |  | #22178 |
|@storybook/addon-docs|@kasperpeulen|✅ | #22180
|@storybook/addon-highlight|@kasperpeulen |✅ | #22181
|@storybook/addon-interactions|@kasperpeulen | |
|@storybook/addon-jest| @Gufah |✅ | #22389 
|@storybook/addon-mdx-gfm| @1234tgk |✅ | #22659
|@storybook/addon-measure| @efrenaragon96 | | #22402
|@storybook/addon-outline| @kyletsang |✅ | #22369 
|@storybook/addon-storyshots| @1234tgk |✅ | #22487 
|@storybook/addon-storyshots-puppeteer| @1234tgk |✅ | #22407
|@storybook/addon-storysource| @1234tgk|✅  | #22367 
|@storybook/addon-viewport| @1234tgk | ✅| #22339
|@storybook/addons| @Beadko  | |
|@storybook/angular| @1234tgk | |
|@storybook/api| | |
|@storybook/blocks| @ssingh3856  | |
|@storybook/channel-postmessage| @EDuToit  | ✅ | https://github.com/storybookjs/storybook/pull/22335
|@storybook/channel-websocket| @EDuToit  |✅ | #22364 
|@storybook/channels| @EDuToit | ✅| #22365
|@storybook/cli| @0916dhkim | | #22254
|@storybook/client-api| @efrenaragon96 |✅  | #22421 
|@storybook/codemod| | |
|@storybook/components| @Beadko | | #22569 
|@storybook/core-client| @kolife01 |✅ | #22447 
|@storybook/core-events| @kolife01 | ✅| #22447
|@storybook/core-server| @stilt0n | ✅ | #23182
|@storybook/csf-tools| @usrrname  |✅  | #22312
|@storybook/docs-tools| @efrenaragon96 | | #22567 
|@storybook/external-docs| | |
|@storybook/html-vite| @usrrname | ✅  | #22293
|@storybook/instrumenter| @kyletsang  | ✅  |#22370
|@storybook/manager| @Dobariya-Nishant  | |
|@storybook/manager-api| @subhajit20 | |
|@storybook/postinstall| @kuriacka |✅  |https://github.com/storybookjs/storybook/pull/22200
|@storybook/preact-vite| @mariasimo| |
|@storybook/preset-create-react-app|@kuriacka | |
|@storybook/preset-vue-webpack| @kyletsang  |✅  | #22320
|@storybook/preset-vue3-webpack|@chakAs3 | ✅ | #23377
|@storybook/react-vite| @mariasimo| ✅  | #22428 
|@storybook/router| @kuriacka  |✅ |https://github.com/storybookjs/storybook/pull/22200
|@storybook/scripts| | |
|@storybook/server| @unpunnyfuns | |
|@storybook/source-loader| @usrrname |✅  | #22420 
|@storybook/svelte-vite| @kolife01 |✅  | #22411 
|@storybook/sveltekit| @kolife01  |✅  | #22412 
|@storybook/theming| @kuriacka  | ✅ | #22376
|@storybook/types| @kuriacka | ✅ | #22397 
|@storybook/vue3-vite|@chakAs3 | ✅ | #23375
|@storybook/vue3-webpack5|@chakAs3 |  ✅| #23376
|@storybook/web-components| @efrenaragon96 | | #22399
|@storybook/web-components-vite| @usrrname| ✅  | #22309

So if you'd like to help converting any of these ♥️ ! Please mention here, which ones, you can take on, and open a PR migrating a package.

If you need any help, please reach out to storybook maintainers on the [contributing channel](https://discord.gg/5pVUdF8gjj) on the Storybook discord.

**Comment by kuriacka** [comment]
Hi,
I would like to help. Some packages like **postinstall** or **router** don't require any TS changes (except changing `tsconfig.json` of course). Do you want to create separate PR for each package or can I merge them into one?

Thanks in advance for your reply.

**Comment by kasperpeulen** [comment]
@kuriacka Sounds great, one PR or multiple are both fine! 

Let me now if you start work on any, then I will put your in the assign list so that me (and others) won't do the same work!

**Comment by kuriacka** [comment]
@kasperpeulen Thanks for info, I can start with aforementioned **postinstall** and **router**, the I can pick another packages.

**Comment by kuriacka** [comment]
@kasperpeulen Here is the [PR for **postinstall** and **router**](https://github.com/storybookjs/storybook/pull/22200), sorry it took so long, but the CI was broken until [fix was merged](https://github.com/storybookjs/storybook/pull/22201). 

I can take another packages, please assign to me **@storybook/theming** and **@storybook/preset-create-react-app**.

Thanks!

**Comment by mariasimo** [comment]
Hey @kasperpeulen! I'd like to help too, but I'm not very familiar yet with all the Storybook packages. Which one do you think can be a good one to start?

Thanks!

**Comment by suiux** [comment]
Hi @kasperpeulen! I'd like to help too!
I don't have much experience with the Storybook packages yet. Could you recommend a good one for me to begin with?
Thanks!

**Comment by kasperpeulen** [comment]
@mariasimo @suiux I think the packages in the frameworks directory are probably the easiest!
So @storybook/web-components-vite, @storybook/react-vite, @storybook/html-vite, @storybook/preact-vite

**Comment by mariasimo** [comment]
@kasperpeulen thanks! I can take ` @storybook/react-vite` and ` @storybook/preact-vite` to begin with, if that's ok!

**Comment by kasperpeulen** [comment]
Sounds great @mariasimo, I assigned them to you :)

**Comment by shilman** [comment]
w00t!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.9 containing PR #22181 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

Closing this issue. Please re-open if you think there's still more to do.

**Comment by 0916dhkim** [comment]
Wait, isn't there a lot more packages that aren't migrated to strict mode yet? I wanted to contribute to @storybook/cli

**Comment by joe-vaughan** [comment]
Apologies! An automation went rogue 😅

Also, for anyone contributing to this or interested in taking part: we've just set up a new Discord working group for the project! Share updates on how you're getting, ask any questions, or simply chat with contributors 🎀

https://discord.com/channels/486522875931656193/1100689736022106173

**Comment by usrrname** [comment]
I'll take `@storybook/html-vite` if no one's on it already

**Comment by unpunnyfuns** [comment]
I'm looking very intently at `@storybook/server`.

**Comment by usrrname** [comment]
Took a stab at `@storybook/csf-tools`... can anyone tell I'm restless on staycation? 🤣

**Comment by EDuToit** [comment]
HI I looked into @storybook/channel-postmessage #22335

**Comment by 1234tgk** [comment]
Hello.

I would like to take a jab at @storybook/addon-viewport. Can you assign it to me, please?

**Comment by kuriacka** [comment]
@kasperpeulen Can I take `@storybook/preset-create-react-app` and `@storybook/theming` now? Thanks.

**Comment by vanessayuenn** [comment]
@kuriacka I assigned those to you. thank you!

**Comment by 1234tgk** [comment]
I just opened the PR for converting @storybook/addon-viewport to strict TS. I think I finished it.
https://github.com/storybookjs/storybook/pull/22339

It passed both `yarn check` and `yarn task --task check --no-link --start-from=install`

Am I missing other tests?

**Comment by shilman** [comment]
Yo-ho-ho!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.12 containing PR #22309 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

Closing this issue. Please re-open if you think there's still more to do.

**Comment by EDuToit** [comment]
I looked into @storybook/channel-websocket #22364

**Comment by 1234tgk** [comment]
@kasperpeulen I am just waiting on [my previous PR #22339](https://github.com/storybookjs/storybook/pull/22339) to be reviewed. In the meanwhile, can I take these packages, please?

@storybook/addon-storyshots
@storybook/addon-storyshots-puppeteer
@storybook/addon-storysource

**Comment by kasperpeulen** [comment]
@1234tgk Reviewed and I will assign you!

**Comment by kasperpeulen** [comment]
Reviewing your stuff @EDuToit Thanks!

**Comment by joe-vaughan** [comment]
That's weird @cyrfer - could you try this link instead?

https://discord.gg/storybook

**Comment by efrenaragon96** [comment]
I am getting `@storybook/addon-measure` too.

**Comment by Beadko** [comment]
Hi, may I take @storybook/components?

**Comment by vanessayuenn** [comment]
@efrenaragon96 @Beadko assigned! thank you for your contribution 🙏🏼

**Comment by kolife01** [comment]
Hi, May I take `@storybook/svelte-vite` and `@storybook/sveltekit` ?  Thanks!

**Comment by shilman** [comment]
Whoopee!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.13 containing PR #22339 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

**Comment by efrenaragon96** [comment]
Hello, may I take `@storybook/cli` and `@storybook/client-api`?

**Comment by usrrname** [comment]
Also took `@storybook/source-loader`

**Comment by kolife01** [comment]
Also took  `@storybook/core-client`  and `@storybook/core-events` .

**Comment by kasperpeulen** [comment]
Hi all, I am on holiday until the 19th of May.
Feel free to start working on any PR! I will review them all after my holiday 🙂

**Comment by shilman** [comment]
Yo-ho-ho!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.14 containing PR #22367 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

**Comment by 1234tgk** [comment]
Hello. These are the PRs for @storybook/addon-storyshots-puppeteer and @storybook/addon-storyshots strict TS migration:
https://github.com/storybookjs/storybook/pull/22487
https://github.com/storybookjs/storybook/pull/22407

While these PRs get reviewed, can I take @storybook/addon-outline and @storybook/addon-mdx-gfm, please?

**Comment by 0916dhkim** [comment]
I forgot to mention here, but I was working on storybook/cli.
https://github.com/storybookjs/storybook/pull/22254
@efrenaragon96 Is it okay if I continue working on that package, unless you have already made good progress?

**Comment by efrenaragon96** [comment]
> I forgot to mention here, but I was working on storybook/cli. #22254 @efrenaragon96 Is it okay if I continue working on that package, unless you have already made good progress?

No problem @0916dhkim I wasn't seen your PR, sorry. I haven't made more progress since I open it so you can continue working on that package, I will close my PR don't worry. Sorry for the confusion.
Please update the table if it is possible @vanessayuenn, thanks

**Comment by efrenaragon96** [comment]
Then I am going to work on `@storybook/docs-tools` if there are no objections, thanks.

**Comment by Beadko** [comment]
PR for the storybook/components https://github.com/storybookjs/storybook/pull/22569

**Comment by Beadko** [comment]
May I pick up the @storybook/addons if it's still free?

**Comment by ssingh3856** [comment]
Is @storybook/blocks still open or is it taken up by anybody else? If not can i work on it?

**Comment by 1234tgk** [comment]
@storybook/addons-mdx-gfm is finished. I just had to switch `strict: false` to `strict: true` https://github.com/storybookjs/storybook/pull/22659

@storybook/addons-outline looks like it already has strict TS. Let me know if that is not the case.

May I take @storybook/angular if no one took it, please?

**Comment by shilman** [comment]
ZOMG!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.1.0-alpha.27 containing PR #22399 that references this issue. Upgrade today to the `@future` NPM tag to try it out!

```
npx sb@next upgrade --tag future
```

**Comment by dlarroder** [comment]
May I take @storybook/manager-api please?

**Comment by mateoguzmana** [comment]
I'm looking into @storybook/manager and @storybook/manager-api as I see those are still free

**Comment by mateoguzmana** [comment]
I'm looking into @storybook/manager and @storybook/manager-api as I see those are still free

**Comment by cupton15** [comment]
Could I take @storybook/codemod?

**Comment by stilt0n** [comment]
I've started working on @storybook/core-server since it appears to be free.

**Comment by aditya1** [comment]
I would like to help with @storybook/codemod

**Comment by Dobariya-Nishant** [comment]
i want to help in @storybook/manager

**Comment by cupton15** [comment]
Ok looks like I didn't finish @storybook/codemod in time, I'll pick up @storybook/core-server

**Comment by stilt0n** [comment]
Hi @cupton15 - I already have a PR open for core-server.

**Comment by mastrzyz** [comment]
Trying to help out with some of the post migration cleanup -> 

https://github.com/storybookjs/storybook/pull/23931
https://github.com/storybookjs/storybook/pull/23932

**Comment by LuisChiej** [comment]
Hello @kasperpeulen, I'm trying to migrate `@storybook/external-docs` but I keep running into errors running `yarn check`
<img width="813" alt="Screenshot 2023-08-23 at 11 50 42" src="https://github.com/storybookjs/storybook/assets/54555805/19b56696-745c-4a37-a7eb-89f5202f36f8">
On running `yarn install`, I get this 
<img width="823" alt="Screenshot 2023-08-23 at 11 43 44" src="https://github.com/storybookjs/storybook/assets/54555805/1aa16950-9fc3-4f1a-b3d5-93384f886f8f">

**Comment by kasperpeulen** [comment]
@LuisChiej I think we can actually skip external docs. I think this is a package that we likely want to remove in the feature, maybe @shilman has more context about that.

**Comment by subhajit20** [comment]
@kasperpeulen  hey I would like to help. Following @storybook/manager-api is still free. can you assign it to me ?

**Comment by kasperpeulen** [comment]
@subhajit20 Done!

**Comment by johnhunter** [comment]
Hey, I'd like to help out with this work. Of the remaining unassigned packages it looks like codemod and scripts are the ones left to do:

- @storybook/api (deprecated, see @storybook/manager-api)
- @storybook/external-docs (going to be deprecated)
- @storybook/codemod
- @storybook/scripts

I can look at either, @kasperpeulen any preference?
